### PR TITLE
WIP: Implement RFC 7440

### PR DESCRIPTION
**Pull request by lzaoral**

### CODE REVIEW

File: tftpd/tftpd.c [hunk 1]
```diff
@@ -1654,16 +1689,49 @@ static void tftp_sendfile(const struct formats *pf, struct tftphdr *oap, int oac
...
+                    break;
+                }
+
+                if ((ap_block < block && block - ap_block < windowsize) ||
+                    (ap_block > block && (block + 65535) - ap_block < windowsize)) {
+                    /* The offset is moved forward only by multiples of segsize */
+                    int acked = 0;
+
+                    /* No-rollover */
+                    if (block - ap_block < block)
```
Comment (lnykryn) [inline tftpd/tftpd.c:1705]: This line looks fishy to me since it might be simplified to "ap_block > 0"

File: tftpd/tftpd.c [hunk 2]
```diff
@@ -1654,16 +1689,49 @@ static void tftp_sendfile(const struct formats *pf, struct tftphdr *oap, int oac
...
             if (ap_opcode == ACK) {
+                /* Whole window sent */
                 if (ap_block == block) {
+                    offset = window = 0;
+                    timeout = rexmtval;
+                    break;
+                }
+
+                if ((ap_block < block && block - ap_block < windowsize) ||
+                    (ap_block > block && (block + 65535) - ap_block < windowsize)) {
```
Comment (lnykryn) [inline tftpd/tftpd.c:1700]: I feel that rollover_val should be used somehow here.

File: tftpd/tftpd.c [hunk 3]
```diff
@@ -1631,16 +1651,31 @@ static void tftp_sendfile(const struct formats *pf, struct tftphdr *oap, int oac
...
             syslog(LOG_WARNING, "tftpd: write: %m");
             goto abort;
         }
-        read_ahead(file, pf->f_convert);
-        for (;;) {
+
+        // TODO: make this useful
+        // read_ahead(file, pf->f_convert);
+
+        while (window == windowsize || size != segsize) {
```
Comment (lnykryn) [inline tftpd/tftpd.c:1678]: This is ugly. Basically, this condition is related to the previous code. I would do something in the term of
for(window=0;window<windowsize;window++) {
send data
}
for(;;;) {
wait and parse the reply
}

File: tftpd/tftpd.c (hunk 1 above)
Comment (lnykryn) [reply tftpd/tftpd.c:1705]: I think this should be ap_block < block

File: tftpd/tftpd.c [hunk 4]
```diff
@@ -1586,7 +1603,8 @@ static void tftp_sendfile(const struct formats *pf, struct tftphdr *oap, int oac
     static u_short block = 1;   /* Static to avoid longjmp funnies */
     u_short ap_opcode, ap_block;
     unsigned long r_timeout;
-    int size, n;
+    int size, n, window;
+    off_t offset = 0;
```
Comment (lnykryn) [inline tftpd/tftpd.c:1607]: Shouldn't this be static?

File: tftpd/tftpd.c [hunk 5]
```diff
@@ -1654,16 +1689,49 @@ static void tftp_sendfile(const struct formats *pf, struct tftphdr *oap, int oac
...
 
             if (ap_opcode == ACK) {
+                /* Whole window sent */
                 if (ap_block == block) {
+                    offset = window = 0;
+                    timeout = rexmtval;
+                    break;
+                }
+
+                if ((ap_block < block && block - ap_block < windowsize) ||
```
Comment (lnykryn) [inline tftpd/tftpd.c:1699]: To make this a bit more readable, I would negate this check and jump to to "fail section in the end"
if(ap obviously wrong)
goto fail;

process the correct ap;

fail:
synchnet......

**Comment by lzaoral** [comment]
Thnak you so much for the initial review, @lnykryn!  Unfortunately, the code was still buggy and did not work properly in some corner cases. :/ This PR is now obsoleted by #2.

**Pull request closed without merging**